### DISCUSSION TITLE
Is there anyway to draw each character by directly rotating counterclockwise 90 degree?

### DISCUSSION THREAD

**Discussion by nissansz** [python-pillow/Pillow#8517] [Q&A] [answered]
Is there anyway to draw each character by directly rotating counterclockwise 90 degree?

![word_5](https://github.com/user-attachments/assets/5accace0-a2f6-430d-86be-a30b97827caa)

**Comment by radarhere** [comment, marked as answer]
Pillow doesn't offer a single API to draw rotated text. For an example of drawing text onto a temporary image, rotating that image, and then pasting it onto another, see https://github.com/python-pillow/Pillow/issues/5738#issuecomment-932834769

**Comment by nissansz** [comment]
Thank you.
For a text line like 10 characters, which is faster?
1. Draw and paste each character.
2. Or Draw whole line vertically, then rotate whole image?

**Reply by radarhere** [reply]
Running a test, yes, a whole line would be faster.

**Comment by nissansz** [comment]
Thank you. But a whole is limited to whole line angle only.
Character angle seems better.

**Comment by nissansz** [comment]
How to update this line to avoid error?
    txt = Image.new(image.mode, font.getsize(text))
AttributeError: 'FreeTypeFontFamily' object has no attribute 'getsize'

```python
def draw_rotated_text(image, font, text, angle, x, y):
    txt = Image.new(image.mode, font.getsize(text))
    d = ImageDraw.Draw(txt)
    d.text((0, 0), text, font=font, fill=(255, 0, 0))
    txt = txt.rotate(angle, expand=1)
    image.paste(txt, (int(x - txt.width/2), int(y - txt.height/2)), txt)
```

**Reply by radarhere** [reply]
I wrote
```python
txt = Image.new(image.mode, font.getsize(text))
```
before `getsize()` was deprecated and later removed.

You can now use
```python
left, top, right, bottom = font.getbbox(text)
width, height = right - left, bottom - top
txt = Image.new(image.mode, (width, height))
```

See https://pillow.readthedocs.io/en/stable/deprecations.html#font-size-and-offset-methods for more information.

**Comment by nissansz** [comment]
Thank you. Another error poped here. 
```python
   d.text((0, 0), text, font=font, fill=(255, 0, 0))
```
```
  File "20241101.py", line 14, in draw_rotated_text
    d.text((0, 0), text, font=font, fill=(255, 0, 0))
  File "D:\Python3.8.8\lib\site-packages\PIL\ImageDraw.py", line 576, in text
    ink = getink(fill)
  File "D:\Python3.8.8\lib\site-packages\PIL\ImageDraw.py", line 514, in getink
    ink, fill = self._getink(fill)
  File "D:\Python3.8.8\lib\site-packages\PIL\ImageDraw.py", line 141, in _getink
    ink = self.draw.draw_ink(ink)
TypeError: color must be int or single-element tuple
```

**Reply by radarhere** [reply]
In the example code, I was drawing red text - `(255, 0, 0)`.

The error is saying that you are trying to draw onto a single band image - either `1` or `L` mode. So you cannot draw red. You need to use a single integer (or a tuple of a single integer) as the fill.

**Comment by nissansz** [comment]
Thank you.
The Chinese character looks fine. English text seems incorrect.
```python
draw_rotated_text(image,font_family, '我', 0, 0+font_size/2, image.height/2)
draw_rotated_text(image, font_family, "们", 90, 0+font_size/2+font_size*1, image.height/2)
draw_rotated_text(image, font_family, "e", 120, 0+font_size/2+font_size*2, image.height/2)
```
![image](https://github.com/user-attachments/assets/0504d7d0-a54a-4863-ab1b-b459334379d6)

**Reply by radarhere** [reply]
Please provide a simple self-contained snippet to demonstrate the problem. Without that, there is insufficient information.

For example, I don't know what font you're using.

**Comment by nissansz** [comment]
I editted your original code, and now another error poped.
```
    image.paste(txt, (int(x - txt.width/2), int(y - txt.height/2)), txt)
  File "D:\Python3.8.8\lib\site-packages\PIL\Image.py", line 1760, in paste
    self.im.paste(im, box, mask.im)
```
```python
from PIL import Image, ImageDraw, ImageFont, ImageFilter


def draw_rotated_text(image, font, text, angle, x, y):
    # txt = Image.new(image.mode, font.getsize(text))

    left, top, right, bottom = font.getbbox(text)
    width, height = right - left, bottom - top
    txt = Image.new(image.mode, (width, height))

    d = ImageDraw.Draw(txt)
    d.text((0, 0), text, font=font, fill=(255, 0, 0))
    # d.text((0, 0), text, font=font, fill=255)
    txt = txt.rotate(angle, expand=1)
    image.paste(txt, (int(x - txt.width/2), int(y - txt.height/2)), txt)



def image_text():
    image = Image.open(r'F:\bg1.jpg')
    # font = ImageFont.truetype('demo.ttf', 50)

    font_size = 50
    font0 = ImageFont.truetype(r'F:\fonts\Arial Unicode MS.ttf', size=font_size - 5)

    fontpath1 = r"F:\fonts\NotoOldShapeNormal.TTF"

    font1 = ImageFont.truetype(fontpath1, size=font_size - 5)
    # font_family = ImageFont.FreeTypeFontFamily(font0, font1, font2)
    font_family = ImageFont.FreeTypeFontFamily(font0, font1)

    draw_rotated_text(image,font_family, '我', 30, 0+font_size/2, image.height/2)
    draw_rotated_text(image, font_family, "们", 50, 0+font_size/2+font_size*1, image.height/2)
    draw_rotated_text(image, font_family, "e", 80, 0+font_size/2+font_size*2, image.height/2)
    image.show()

if __name__ == '__main__':
    image_text()
```

**Reply by radarhere** [reply]
bg1.jpg is an RGB image. The error is because
```python
txt = Image.new(image.mode, (width, height))
```
creates a new RGB image, but without transparency, the paste operation is failing to remove the non-letter parts of the image.

You can use
```python
txt = Image.new("RGBA", (width, height))
```
instead.

**Comment by nissansz** [comment]
![bg1](https://github.com/user-attachments/assets/72adfe5a-f259-4a88-8c9b-5e31267e7c8c)
above is bg1.jpg

**Comment by nissansz** [comment]
[fonts.zip](https://github.com/user-attachments/files/17599349/fonts.zip)
Above are 2 fonts tested

**Comment by nissansz** [comment]
Ok. Result is below, seems half cut for Chinese characters, and missing English letter.

![image](https://github.com/user-attachments/assets/9cf51598-6b71-439e-9143-b63801b5e8d9)

**Reply by radarhere** [reply]
Ah. I didn't consider the top and left offset when transitioning to `getbbox()`. There is a gap surrounding the character when it is drawn, which isn't helpful if you just want to rotate the drawn pixels.

Replace
```python
left, top, right, bottom = font.getbbox(text)
width, height = right - left, bottom - top
txt = Image.new(image.mode, (width, height))

d = ImageDraw.Draw(txt)
d.text((0, 0), text, font=font, fill=(255, 0, 0))
```
with
```python
left, top, right, bottom = font.getbbox(text)
txt = Image.new("RGBA", (right, bottom))

d = ImageDraw.Draw(txt)
d.text((0, 0), text, font=font, fill=(255, 0, 0))
txt = txt.crop((left, top, right, bottom))
```

**Comment by nissansz** [comment]
You are too strong.
It works perfect.
![image](https://github.com/user-attachments/assets/d0860c1a-f6f0-4ec9-8c1c-98cc68879b67)
